From 9e7cfb617f8c66bf0e06caed2a1c8c97d5e611fd Mon Sep 17 00:00:00 2001
From: Ben Levinsky <ben.levinsky@xilinx.com>
Date: Tue, 31 Mar 2020 10:58:30 -0700
Subject: [PATCH 1306/1852] remoteproc: zynqmp_r5: Add explicit nodes for
 vrings

commit 5f6e1f320fefac242e9581590298348ca1335f16 from
https://github.com/Xilinx/linux-xlnx.git

Match upstream remoteproc kernel driver convention and add device
tree nodes for vrings. Do this explicitly to ensure appropriate
ranges of memory are used for remoteproc communication.

This patch accompanies linux patch to parse vrings such that if
these nodes are present, then only use memory described with
 these nodes and not from CMA memory.

Signed-off-by: Ben Levinsky <ben.levinsky@xilinx.com>
Acked-by: Wendy Liang <wendy.liang@xilinx.com>
State: not-upstreamable
Signed-off-by: Yaliang Wang <Yaliang.Wang@windriver.com>
---
 drivers/remoteproc/zynqmp_r5_remoteproc.c | 34 +++++++++++++++++++++--
 1 file changed, 31 insertions(+), 3 deletions(-)

diff --git a/drivers/remoteproc/zynqmp_r5_remoteproc.c b/drivers/remoteproc/zynqmp_r5_remoteproc.c
index edb46c8bebb5..4c9103d0c211 100644
--- a/drivers/remoteproc/zynqmp_r5_remoteproc.c
+++ b/drivers/remoteproc/zynqmp_r5_remoteproc.c
@@ -303,7 +303,7 @@ static int zynqmp_r5_parse_fw(struct rproc *rproc, const struct firmware *fw)
 			int id;
 			char name[16];
 
-			id = node->name[8] - 48;
+			id = node->name[8] - '0';
 			snprintf(name, sizeof(name), "vdev%dbuffer", id);
 			/* Register DMA region */
 			mem = rproc_mem_entry_init(dev, NULL,
@@ -312,11 +312,39 @@ static int zynqmp_r5_parse_fw(struct rproc *rproc, const struct firmware *fw)
 						   NULL, NULL,
 						   name);
 			if (!mem) {
-				dev_err(dev, "unable to initialize memory-region %s \n",
-						name);
+				dev_err(dev, "unable to initialize memory-region %s\n",
+						node->name);
 				return -ENOMEM;
 			}
+			dev_dbg(dev, "parsed %s at  %llx\r\n", mem->name,
+				mem->dma);
+			rproc_add_carveout(rproc, mem);
+			continue;
+		} else if (strstr(node->name, "vdev") &&
+				    strstr(node->name, "vring")) {
+			int id, vring_id;
+			char name[16];
 
+			id = node->name[8] - '0';
+			vring_id = node->name[14] - '0';
+			snprintf(name, sizeof(name), "vdev%dvring%d", id,
+				 vring_id);
+			/* Register vring */
+			mem = rproc_mem_entry_init(dev, NULL,
+						   (dma_addr_t)rmem->base,
+						   rmem->size, rmem->base,
+						   NULL, NULL,
+						   name);
+			mem->va = devm_ioremap_wc(dev, rmem->base, rmem->size);
+			if (!mem->va)
+				return -ENOMEM;
+			if (!mem) {
+				dev_err(dev, "unable to initialize memory-region %s\n",
+						node->name);
+				return -ENOMEM;
+			}
+			dev_dbg(dev, "parsed %s at %llx\r\n", mem->name,
+				mem->dma);
 			rproc_add_carveout(rproc, mem);
 			continue;
 		} else {
-- 
2.31.1

